src/meta/processors/partsMan/CreateSpaceProcessor.cpp

```diff
@@ -162,46 +190,113 @@ void CreateSpaceProcessor::process(const cpp2::CreateSpaceReq& req) {
-            return pickedHosts;
+StatusOr<std::unordered_map<HostAddr, int32_t>>
+CreateSpaceProcessor::getHostLoading() {
+    std::unique_ptr<kvstore::KVIterator> iter;
+    const auto& prefix = MetaServiceUtils::partPrefix();
+    auto code = kvstore_->prefix(kDefaultSpaceId, kDefaultPartId, prefix, &iter);
+    if (code != kvstore::ResultCode::SUCCEEDED) {
+        LOG(ERROR) << "List Parts Failed";
+        return Status::Error("List Parts Failed");
+    }
+
+    std::unordered_map<HostAddr, int32_t> result;
+    while (iter->valid()) {
+        auto hosts = MetaServiceUtils::parsePartVal(iter->val());
+        for (auto& host : hosts) {
+            result[host]++;
+        }
+        iter->next();
+    }
+    return result;
+}
+
+StatusOr<Hosts>
+CreateSpaceProcessor::pickHostsWithZone(const std::vector<std::string>& zones,
+                                        std::unordered_map<HostAddr, int32_t>& loading,
+                                        std::unordered_map<std::string, Hosts>& zoneHosts) {
+    Hosts pickedHosts;
+    for (auto iter = zoneHosts.begin(); iter != zoneHosts.end(); iter++) {
+        auto zoneIter = std::find(std::begin(zones), std::end(zones), iter->first);
+        if (zoneIter == std::end(zones)) {
+            continue;
         }
 
-        auto zoneKey = MetaServiceUtils::zoneKey(zoneName);
+        HostAddr picked;
+        int32_t size = INT_MAX;
+        for (auto& host : iter->second) {
+            auto hostIter = loading.find(host);
+            if (hostIter == loading.end()) {
+                LOG(ERROR) << "Host " << host << " not found";
+                handleErrorCode(cpp2::ErrorCode::E_NO_HOSTS);
+                onFinished();
+                return Status::Error("Host not found");
+            }
+
+            if (size > hostIter->second) {
+                picked = host;
+                size = hostIter->second;
+            }
+        }
+
+        loading[picked] += 1;
+        pickedHosts.emplace_back(toThriftHost(std::move(picked)));
+    }
+    return pickedHosts;
+}
+
+StatusOr<std::vector<std::string>>
+CreateSpaceProcessor::pickLightLoadZones(const std::vector<std::string>& zones,
+                                         int32_t replicaFactor,
+                                         std::unordered_map<HostAddr, int32_t>& loading,
+                                         std::unordered_map<std::string, Hosts>& zoneHosts) {
+    std::unordered_map<std::string, int32_t> zoneLoading;
+    for (auto& zone : zones) {
+        auto zoneKey = MetaServiceUtils::zoneKey(zone);
         auto zoneValueRet = doGet(std::move(zoneKey));
         if (!zoneValueRet.ok()) {
-            LOG(ERROR) << "Get zone " << zoneName << " failed";
-            handleErrorCode(cpp2::ErrorCode::E_STORE_FAILURE);
-            onFinished();
-            return pickedHosts;
+            LOG(ERROR) << "Get zone " << zone << " failed";
+            return Status::Error("Get Zone Failed");
         }
```

Comment:
These could move out from `pickLightLoadZones`. No need to read for each partition.